.github/actions/check-changes/action.yaml

```diff
@@ -31,26 +35,42 @@ runs:
       env:
         GIT_BASE_REF: ${{ inputs.baseRef }}
         GIT_REF: ${{ inputs.ref }}
+        GITHUB_EVENT_FORCED: ${{ github.event.forced }}
+        GITHUB_REF_BEFORE: ${{ github.event.before }}
         USE_FALLBACK: ${{ inputs.useFallback }}
+        DIFF_FILTER: ${{ inputs.diffFilter }}
       run: |
         : Check for Changed Files ✅
         if [[ "${RUNNER_DEBUG}" ]]; then set -x; fi
         shopt -s extglob
         shopt -s dotglob
 
-        if ! git cat-file -e "${GIT_BASE_REF}" &> /dev/null; then
-          echo "::warning::Provided base reference ${GIT_BASE_REF} is invalid"
-          if [[ "${USE_FALLBACK}" == 'true' ]]; then
-            GIT_BASE_REF='HEAD~1'
+        if [[ "${GIT_BASE_REF}" ]]; then
+          if ! git cat-file -e "${GIT_BASE_REF}" &> /dev/null; then
+            echo "::warning::Provided base reference ${GIT_BASE_REF} is invalid"
+            if [[ "${USE_FALLBACK}" == 'true' ]]; then
+              GIT_BASE_REF='HEAD~1'
+            fi
           fi
+        else
+          if ! git cat-file -e ${GITHUB_REF_BEFORE} &> /dev/null; then
+            GITHUB_REF_BEFORE='4b825dc642cb6eb9a060e54bf8d69288fbee4904'
+          fi
+
+          GIT_BASE_REF='HEAD~1'
+          case "${GITHUB_EVENT_NAME}" in
+            pull_request) GIT_BASE_REF="origin/${GITHUB_BASE_REF}" ;;
+            push) if [[ "${GITHUB_EVENT_FORCED}" != 'true' ]]; then GIT_BASE_REF="${GITHUB_REF_BEFORE}"; fi ;;
+            *) ;;
+          esac
         fi
 
-        changes=($(git diff --name-only ${GIT_BASE_REF} ${GIT_REF} -- ${{ inputs.checkGlob }}))
+        changes=($(git diff --name-only --diff-filter="${DIFF_FILTER}" ${GIT_BASE_REF} ${GIT_REF} -- ${{ inputs.checkGlob }}))
 
         if (( ${#changes[@]} )); then
           file_string="${changes[*]}"
           echo "hasChangedFiles=true" >> $GITHUB_OUTPUT
-          echo "changedFiles=[${file_string// /,}]" >> GITHUB_OUTPUT
+          echo "changedFiles=[\"${file_string// /\",\"}\"]" >> $GITHUB_OUTPUT
```

Comment:
We were originally calling toJSON on this, but now we are not. What was the intent of the curly brace element?

Reply:
The original idea was to have the output be a valid JSON array with all the changed files. In tests this proved to be insufficient, as the string tokens had to enclosed in quotes, which mean that additional string operations were needed:

So `${file_string// /\",\"}` replaces every space with `\",\"`, so `a b c` becomes `a\",\"b\",\"c`.

This is then enclosed with `\"` again to generate `\"a\",\"b\",\"c\"` which is a properly list of JSON string array tokens.

Reply:
`toJSON` was removed because outputs for composite actions are always of type `string`, so we cannot provide a complex type as an output. Instead we provide a JSON string representation of that type.

Reply:
Ah, I see. It's ~~cursed~~ fancy string substitution in-place before it gets echoed to $GITHUB_OUTPUT. I'd have probably just done the string substitution where I didn't have to escape the quotes, but that may be personal preference.
```yaml
file_string="${file_string// /,}"
echo "changedFiles=[\"${file_string}\"]" >> $GITHUB_OUTPUT
```

Reply:
That's incomplete, you cannot just use a single comma as a glue, you need to use the closing double quotes of the prior token, followed by the comma, and the opening double quotes for the next token again.

In your example you end up with `["a,b,c"]` which would be an array with a single element instead of 3 separate elements.

Reply:
Ah, I see now what the goal is. I'll think on it.

Reply:
To do what I was thinking (separate the string substitution from the final output), you'd have to do this:
```sh
file_string=\"${file_string// /\",\"}\"
echo "changedFiles=[${file_string}]" >> $GITHUB_OUTPUT
```

It doesn't really save any escaping, though it does provide a bit clearer separation of what is occurring at the cost of one-liner elegance. Let me know what you prefer.

Reply:
Again, that would not be valid syntax, because you have to wrap the first string in double quotes again (otherwise string expansion is not applied) and I don't see much benefit in splitting it up, instead we trigger two expansions where one would suffice.

Reply:
I don't see how it's not valid syntax. I tested it repeatedly locally and ensured that it achieves the desired result.

From my perspective, the separation provides a clearer picture of why the escapes are being done. I literally could not read the original syntax in a single pass. It's not like the extra expansion costs us anything measurable here.

That said, I do not want to bike shed this to death, so it's fine as-is.

Reply:
Discussed off-thread and @PatTheMav did say that this does seem to be valid syntax. However, it did occur to me that this action is run in bash and I tested in zsh, so I don't know for sure if it's still valid in bash, though I think it would be valid.

That said, this is still fine as-is, if not a little confusing at first glance.